src/services/GasEstimator.ts

```diff
@@ -0,0 +1,140 @@
+
+export class GasEstimator {
+  public static async apply(
+    provider: ethers.providers.Provider,
+    minGasPrice: number,
+    maxGasPrice: number
+  ): Promise<GasEstimation> {
+    const [block, currentGasPrice, feeData] = await GasEstimator.fetchProviderInfo(provider);
+
+    const [metrics, prices, maxPriorityFees] = GasEstimator.getGasMetricsFor(block);
+
+    return GasEstimator.estimate({
+      currentGasPrice: currentGasPrice.toNumber(),
+      feeData,
+      metrics,
+      prices,
+      minGasPrice,
+      maxGasPrice,
+      maxPriorityFees,
+    });
+  }
+
+  public static printable(metrics: GasEstimation): string {
+    const { gasPrice, maxPriorityFeePerGas, maxFeePerGas, min, max, avg, isTxType2, baseFeePerGas } = metrics;
+
+    return (
+      `isTxType2: ${isTxType2 ? 'yes' : 'no'}, ` +
+      `gasPrice: ${formatWeiToGwei(gasPrice)} Gwei, ` +
+      `baseFeePerGas: ${baseFeePerGas > 1e6 ? formatWeiToGwei(baseFeePerGas) + ' Gwei' : baseFeePerGas}, ` +
+      `maxPriorityFee: ${maxPriorityFeePerGas ? formatWeiToGwei(maxPriorityFeePerGas) : '-'} Gwei, ` +
+      `maxFee: ${maxFeePerGas ? formatWeiToGwei(maxFeePerGas) : '-'} Gwei, ` +
+      `min: ${formatWeiToGwei(min)} Gwei, ` +
+      `max: ${formatWeiToGwei(max)} Gwei, ` +
+      `avg: ${formatWeiToGwei(avg)} Gwei`
+    );
+  }
+
+  private static async fetchProviderInfo(
+    provider: ethers.providers.Provider
+  ): Promise<[BlockWithTransactions, BigNumber, FeeData]> {
+    return Promise.all([provider.getBlockWithTransactions('latest'), provider.getGasPrice(), provider.getFeeData()]);
+  }
+
+  private static getGasMetricsFor(
+    block: BlockWithTransactions
+  ): [gasPriceMetrics: GasEstimation, prices: number[], maxPriorityFees: number[]] {
+    const isTxType2 = Boolean(block.baseFeePerGas);
+    const transactions = block.transactions.filter(GasEstimator.containsGasPrice);
+    const prices = transactions.map(GasEstimator.getGasPrice);
+
+    if (!prices.length) {
+      return [{ min: 0, max: 0, avg: 0, baseFeePerGas: 0, gasPrice: 0, isTxType2 }, [], []];
+    }
+
+    const fees = transactions.filter(GasEstimator.containsFees).map(GasEstimator.getFee);
+    const averageGasPrice = prices.reduce((acc, gas) => acc + gas, 0) / prices.length;
+    const [min, max] = GasEstimator.getMinAndMaxGasPrices(transactions);
+
+    return [
+      {
+        isTxType2,
+        min,
+        max,
+        avg: averageGasPrice,
+        gasPrice: 0,
+        baseFeePerGas: block.baseFeePerGas?.toNumber() || 0,
+      },
+      prices,
+      fees,
+    ];
+  }
+
+  private static estimate(params: EstimateParams): GasEstimation {
+    let baseFee = Math.min(Math.max(params.metrics.baseFeePerGas, params.minGasPrice), params.maxGasPrice);
+
+    if (params.prices.length > 2 && !params.metrics.isTxType2) {
+      baseFee = Type1GasEstimator.calculateBaseGasPrice(params);
+    }
+
+    return GasEstimator.makeGasEstimation(baseFee, params);
```

Comment:
Maybe you can move here the Type2GasEstimator here, just to keep both TypeGasEstimators close

Reply:
In fact, they are pretty close. Look at lines 90-95.
```ts
  private static makeGasEstimation(baseFee: number, params: EstimateParams): GasEstimation {
    if (params.metrics.isTxType2) {
      return Type2GasEstimator.estimate(baseFee, params);
    }
    return Type1GasEstimator.estimate(baseFee, params);
  }
```
But now I understand your comment. I think we might move `baseFee` to its respective estimator so we avoid calculating twice on the same class.

Reply:
Refactored it!